File: git-reword-commit-cli/bindings/dart/lib/src/git_operations.dart
part of '../git_reword.dart';

/// Git operations wrapper.
class GitOperations {
  Pointer<Void>? _handle;

  /// Create a new GitOperations for the given repository path.
  GitOperations([String repoPath = '.']) {
    final pathPtr = repoPath.toNativeUtf8();
    _handle = _NativeBindings.gitOperationsCreate(pathPtr);
    malloc.free(pathPtr);
  }

  /// Check if git is available.
  bool get isGitAvailable {
    _checkHandle();
    return _NativeBindings.gitOperationsIsGitAvailable(_handle!) != 0;
  }

  /// Check if current directory is a git repository.
  bool get isGitRepository {
    _checkHandle();
    return _NativeBindings.gitOperationsIsGitRepository(_handle!) != 0;
  }

  /// Check if working directory is clean.
  bool get isWorkingDirectoryClean {
    _checkHandle();
    return _NativeBindings.gitOperationsIsWorkingDirectoryClean(_handle!) != 0;
  }

  /// Get the HEAD commit hash.
  String? get headCommit {
    _checkHandle();
    final ptr = _NativeBindings.gitOperationsGetHeadCommit(_handle!);
    if (ptr == nullptr) {
      return null;
    }
    final result = ptr.toDartString();
    _NativeBindings.freeString(ptr);
    return result;
  }

  /// Dispose of the resources.
  void dispose() {
    if (_handle != null) {
      _NativeBindings.gitOperationsDestroy(_handle!);
      _handle = null;
    }
  }

  void _checkHandle() {
    if (_handle == null) {
      throw StateError('GitOperations has been disposed');
    }
  }
}

File: git-reword-commit-cli/bindings/java/src/com/gitreword/GitOperations.java
package com.gitreword;

/**
 * Git operations wrapper.
 */
public class GitOperations implements AutoCloseable {
    private long nativeHandle;
    
    static {
        NativeLibrary.load();
    }
    
    public GitOperations() {
        this(".");
    }
    
    public GitOperations(String repoPath) {
        nativeHandle = nativeCreate(repoPath);
    }
    
    public boolean isGitAvailable() {
        checkHandle();
        return nativeIsGitAvailable(nativeHandle);
    }
    
    public boolean isGitRepository() {
        checkHandle();
        return nativeIsGitRepository(nativeHandle);
    }
    
    public boolean isWorkingDirectoryClean() {
        checkHandle();
        return nativeIsWorkingDirectoryClean(nativeHandle);
    }
    
    public String getHeadCommit() {
        checkHandle();
        return nativeGetHeadCommit(nativeHandle);
    }
    
    @Override
    public void close() {
        if (nativeHandle != 0) {
            nativeDestroy(nativeHandle);
            nativeHandle = 0;
        }
    }
    
    private void checkHandle() {
        if (nativeHandle == 0) {
            throw new IllegalStateException("GitOperations has been closed");
        }
    }
    
    // Native methods
    private static native long nativeCreate(String repoPath);
    private static native void nativeDestroy(long handle);
    private static native boolean nativeIsGitAvailable(long handle);
    private static native boolean nativeIsGitRepository(long handle);
    private static native boolean nativeIsWorkingDirectoryClean(long handle);
    private static native String nativeGetHeadCommit(long handle);
}
